src/librustc_trans/base.rs

```diff
@@ -1069,34 +618,62 @@ pub fn trans_ctor_shim<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
     let fn_ty = FnType::new(ccx, Abi::Rust, &sig, &[]);
 
-    let (arena, fcx): (TypedArena<_>, FunctionContext);
-    arena = TypedArena::new();
-    fcx = FunctionContext::new(ccx, llfndecl, fn_ty, None, &arena);
-    let bcx = fcx.init(false);
-
+    let fcx = FunctionContext::new(ccx, llfndecl, fn_ty);
+    let bcx = fcx.get_entry_block();
     if !fcx.fn_ty.ret.is_ignore() {
-        let dest = fcx.llretslotptr.get().unwrap();
+        // But if there are no nested returns, we skip the indirection
+        // and have a single retslot
+        let dest = if fcx.fn_ty.ret.is_indirect() {
+            get_param(fcx.llfn, 0)
+        } else {
+            // We create an alloca to hold a pointer of type `ret.original_ty`
+            // which will hold the pointer to the right alloca which has the
+            // final ret value
+            fcx.alloca(fcx.fn_ty.ret.memory_ty(ccx), "sret_slot")
+        };
         let dest_val = adt::MaybeSizedValue::sized(dest); // Can return unsized value
         let mut llarg_idx = fcx.fn_ty.ret.is_indirect() as usize;
         let mut arg_idx = 0;
         for (i, arg_ty) in sig.inputs().iter().enumerate() {
-            let lldestptr = adt::trans_field_ptr(bcx, sig.output(), dest_val, Disr::from(disr), i);
+            let lldestptr = adt::trans_field_ptr(&bcx, sig.output(), dest_val, Disr::from(disr), i);
             let arg = &fcx.fn_ty.args[arg_idx];
             arg_idx += 1;
-            let b = &bcx.build();
-            if common::type_is_fat_ptr(bcx.tcx(), arg_ty) {
+            if common::type_is_fat_ptr(bcx.ccx(), arg_ty) {
                 let meta = &fcx.fn_ty.args[arg_idx];
                 arg_idx += 1;
-                arg.store_fn_arg(b, &mut llarg_idx, get_dataptr(bcx, lldestptr));
-                meta.store_fn_arg(b, &mut llarg_idx, get_meta(bcx, lldestptr));
+                arg.store_fn_arg(&bcx, &mut llarg_idx, get_dataptr(&bcx, lldestptr));
+                meta.store_fn_arg(&bcx, &mut llarg_idx, get_meta(&bcx, lldestptr));
             } else {
-                arg.store_fn_arg(b, &mut llarg_idx, lldestptr);
+                arg.store_fn_arg(&bcx, &mut llarg_idx, lldestptr);
             }
         }
-        adt::trans_set_discr(bcx, sig.output(), dest, disr);
-    }
+        adt::trans_set_discr(&bcx, sig.output(), dest, disr);
+
+        if fcx.fn_ty.ret.is_indirect() {
+            bcx.ret_void();
+            return;
+        }
 
-    fcx.finish(bcx, DebugLoc::None);
+        if let Some(cast_ty) = fcx.fn_ty.ret.cast {
+            let load = bcx.load(bcx.pointercast(dest, cast_ty.ptr_to()));
+            let llalign = llalign_of_min(fcx.ccx, fcx.fn_ty.ret.ty);
+            unsafe {
+                llvm::LLVMSetAlignment(load, llalign);
+            }
+            bcx.ret(load)
+        } else {
+            let llty = fcx.fn_ty.ret.original_ty;
+            let retval = if llty == Type::i1(fcx.ccx) {
```

Comment:
This still needs to be removed. Also, I wonder if `abi::ArgTy` has any method to do this sort of thing to get you the value that you can pass to `bcx.ret` (`fn_ty.ret_ty` is an `ArgTy`).